pkg/dyninst/decode/marshal.go

```diff
@@ -32,6 +37,36 @@ type debuggerData struct {
 	EvaluationErrors []string     `json:"evaluationErrors,omitempty"`
 }
 
+type message struct {
+	probe        *ir.Probe
+	captureEvent *captureEvent
+}
+
+func (m *message) MarshalJSONTo(enc *jsontext.Encoder) error {
+	var sb strings.Builder
+
+	// Go through each template segment, if its a string, write it directly to the encoder, if its a json segment, process the expression
+	// and write the result to the encoder.
+	for _, seg := range m.probe.Template.Segments {
+		switch segTyped := seg.(type) {
+		case ir.JSONSegment:
+			// Extract raw value from expression (no JSON encoding)
+			value, err := m.captureEvent.extractExpressionRawValue(segTyped.ExpressionIndex)
```

Comment:
How does this interact with evaluation errors?

Reply:
I'm glad you pointed this out - at the time of your comment the answer is "It doesn't". 

I'm pushing a commit now which changes error handling in `message.MarshalJSONTo` such that evaluation of each individual segment in a probe's template that causes an error will append to `evaluationErrors`. I'm going to add tests in a follow up commit which exercises this.

Reply:
Pushed tests (same commit)